src/v/cluster/log_eviction_stm.cc

```diff
@@ -9,104 +9,345 @@
+            _storage_eviction_offset = std::min(
+              last_evicted, max_collectible_offset);
+            _reap_condition.signal();
+        } catch (const ss::abort_requested_exception&) {
+            // ignore abort requested exception, shutting down
+        } catch (const ss::gate_closed_exception&) {
+            // ignore gate closed exception, shutting down
+        } catch (const std::exception& e) {
+            vlog(
+              _logger.info,
+              "Error handling log eviction - {}, ntp: {}",
+              e,
+              _raft->ntp());
+        }
+    }
+}
 
-    model::offset max_collectible_offset
-      = _stm_manager->max_collectible_offset();
-    if (last_evicted > max_collectible_offset) {
+ss::future<> log_eviction_stm::do_write_raft_snapshot(
+  model::offset nearest_segment_boundary) {
+    co_await _raft->visible_offset_monitor().wait(
+      _delete_records_eviction_offset, model::no_timeout, _as);
+    co_await _raft->refresh_commit_index();
+    co_await _raft->log().stm_manager()->ensure_snapshot_exists(
+      nearest_segment_boundary);
+    const auto max_collectible_offset
+      = _raft->log().stm_manager()->max_collectible_offset();
+    if (nearest_segment_boundary <= max_collectible_offset) {
+        vlog(
+          _logger.debug,
+          "Reaping data before offset: {} - for ntp: {}",
+          nearest_segment_boundary,
+          _c->ntp());
+        co_await _raft->write_snapshot(
+          raft::write_snapshot_cfg(nearest_segment_boundary, iobuf()));
+    } else {
         vlog(
           _logger.trace,
-          "Can only evict up to offset: {}, ntp: {}",
-          max_collectible_offset,
+          "Cannot prefix truncate due to ongoing processing on data "
+          "below the requested new start offset {} for ntp {}, "
+          "retrying... ",
+          nearest_segment_boundary,
           _raft->ntp());
-        last_evicted = max_collectible_offset;
+    }
+}
+
+model::offset log_eviction_stm::effective_start_offset() const {
+    /// The start offset is the max of either the last snapshot index or the
+    /// most recent delete records eviciton offset. This is because even after
+    /// bootstrap the larger of the two will reflect the most recent event that
+    /// has occurred, and will be the correct start offset.
+    ///
+    /// NOTE: Cannot replace last_snapshot_index with _storage_eviction_offset
+    /// as this is the requested eviction offset and its not persisted anywhere.
+    /// In the event this is set but a crash occurred before write_snapshot was
+    /// called (occurs in background) it would appear that the start offset was
+    /// incremented then returned to a previous value.
+    return model::next_offset(
+      std::max(_raft->last_snapshot_index(), _delete_records_eviction_offset));
+}
+
+ss::future<std::error_code> log_eviction_stm::truncate(
+  model::offset kafka_truncate_offset,
```

Comment:
would it be possible for us to perform the translation in `kafka` layer i.e. replicated_partition, we generally stick to do that and only exceptionally do it in other places. In this case it seems that we can use translated offset here

Reply:
This is sometime during the handling of the request, so maybe it would belong in `kafka/handler/delete-records.cc` or `cluster/partition.cc`?

Reply:
I suspect we will run into trouble doing it in the Kafka layer, given this stm needs to know about the kafka offset to pass it to the archival layer, which isn't able to do offset translation. I.e., this call needs to work with an offset that doesn't exist in local data.